README.md
<div align="center">

![Logo](https://cdn.discordapp.com/attachments/892836872118763543/1179125566360793251/HowStrongIsIt-Logo.png?ex=6578a542&is=65663042&hm=7d853bbe01a769f7ce552f9c57f7701fbc7832970b2d2fcda704a5d1e487a19c&)

[![License](https://img.shields.io/badge/License-MIT-blue.svg?style=flat-square)](https://github.com/inttter/HowStrongIsIt/blob/main/LICENSE) 
[![Last Commit](https://img.shields.io/github/last-commit/inttter/HowStrongIsIt?style=flat-square)](https://github.com/inttter/HowStrongIsIt/commits/main)
[![Website Status](https://img.shields.io/website?url=https%3A%2F%2Finttter.github.io%2FHowStrongIsIt%2F)](https://inttter.github.io/HowStrongIsIt/)
[![CodeFactor](https://www.codefactor.io/repository/github/inttter/howstrongisit/badge)](https://www.codefactor.io/repository/github/inttter/howstrongisit)
![Open Issues](https://img.shields.io/github/issues/inttter/HowStrongIsIt?style=flat-square)
![Watchers](https://img.shields.io/github/watchers/inttter/HowStrongIsIt?style=flat-square)

<a href="https://www.buymeacoffee.com/intter" target="_blank"><img src="https://cdn.buymeacoffee.com/buttons/v2/default-blue.png" alt="Buy Me A Coffee" style="height: 40px !important;width: 150px !important;" ></a>

# 🔒 **HowStrongIsIt** 🔒

Uncertain about your password strength? This website lets you check to see how strong (or weak...) your password is.

## 🪷 Features 🪷

- **Password Checker** - Enter your password and check it's strength. No extra steps. No logging data.

- **Tooltips** - Hover over the tooltip in the top left for a general rundown of how a good and bad password looks like.

- **General Password Practises** - Don't know how to keep your password secure? Click on *'general password practises!'* to see 5 tips on how to stay secure and not get your accounts hacked.

## 🍫 Cloning the repository? 🍫

Follow these steps:

1. Clone the repository:

`git clone https://github.com/inttter/HowStrongIsIt.git`

2. Open `index.html` in your browser.

## 👨‍💻 Technologies Used 👨‍💻

     HTML
     CSS
     JavaScript

## 🥝 Contributing 🥝

Want to contribute? Fork the repository and submit a pull request!

## 🏆 Acknowledgments 🏆

[Brad Traversy and his 50Projects50Days project](https://github.com/bradtraversy/50projects50days/tree/master/todo-list) for the general foundations of this website I used.

[FontAwesome](https://fontawesome.com) for the Dark Mode toggle icon.

## ⚖️ License ⚖️

[HowStrongIsIt](https://github.com/inttter/HowStrongIsIt) is licensed under the MIT License.

See the [LICENSE](LICENSE) file for more information.

## ☕ Buy Me A Coffee! ☕

<a href="https://www.buymeacoffee.com/intter" target="_blank"><img src="https://cdn.buymeacoffee.com/buttons/v2/default-blue.png" alt="Buy Me A Coffee" style="height: 40px !important;width: 150px !important;" ></a>

CHANGELOG.md
26/11/23: 

**V1.0**

    - Initial functionality for the website


27/11/23: 

**V1.1**

✅ - Created a HTML page for general security practises
    
    - Added multiple secure password tips
    
    - Added a button on the main page that links to the page

    - Completed styling to the page

    - Added a custom scrollbar

    - Added animations to some elements on the page

    - Added some widgets

✅ - Added messages for repeating patterns (like 123, ABC, 246, etc)

✅ - Added a tooltip for what a reccomended password should look like.

✅ - Added more options for how strong a password is

✅ - Added more buttons


✅ - Added a toggle for dark mode

✅ - Styled the page more

✅ - Added animations to most elements on the page!

✅ - Added widgets to other HTML pages

---
✅ - Created a HTML page for session timeout



    - If the user is inactive for 30 seconds, the page will load a HTML page "timeout.html" which displays a message that the page has been closed for security reasons, and provides a redirect link back to the main page

✅ - Mostly the same as the last page

28/11/23:

**V1.2**

✅ - Created an extensive README

✅ - Made a markdown file which has the contents of recommended.html + a little bit more detail and/or images.

✅ - Slight adjustments

✅ - Added reference to PASSWORD-PRACTISES.md on recommended.html page


script.js
const passwordInput = document.getElementById('passwordInput');
const passwordStrength = document.getElementById('passwordStrength');


passwordInput.addEventListener('input', function () {
  const password = passwordInput.value;
  const strength = checkPasswordStrength(password);
  displayPasswordStrength(strength);
});

function checkPasswordStrength(password) {
  const patterns = {
    repeatingChars: /(.)\1{2,}/, // Pattern to check for repeating characters
    sequentialNumbers: /123|234|345|456|567|678|789|987|876|765|654|543|432|321/g, 
    sequentialLetters: /abc|bcd|cde|def|efg|fgh|ghi|hij|ijk|jkl|klm|lmn|mno|nop|opq|pqr|qrs|rst|stu|tuv|uvw|vwx|wxy|xyz|zyx|yxw|xwv|wvu|vut|uts|tsr|srq|rqp|qpo|pon|onm|nml|mlk|lkj|kji|jih|ihg|hgf|gfe|fed|edc|dcb|cba/gi
  };

  if (password.length >= 20 && /[a-z]/.test(password) && /[A-Z]/.test(password) && /[0-9]/.test(password) && /[!@#$%^&*()_+\-=[\]{};':"\\|,.<>/?]/.test(password)) {
    return 'unbreakable';
  } else if (password.length >= 16) {
    if (patterns.repeatingChars.test(password)) {
      return "don't use repeating characters!";
    } else if (patterns.sequentialNumbers.test(password) || patterns.sequentialLetters.test(password)) {
      return "don't use sequences like this!";
    } else {
      return 'very strong';
    }
  } else if (password.length >= 12) {
    if (patterns.repeatingChars.test(password)) {
      return "don't use repeating characters!";
    } else if (patterns.sequentialNumbers.test(password) || patterns.sequentialLetters.test(password)) {
      return "don't use sequences like this!";
    } else {
      return 'strong';
    }
  } else if (password.length >= 8) {
    if (patterns.repeatingChars.test(password)) {
      return "Don't use repeating characters!";
    } else if (patterns.sequentialNumbers.test(password) || patterns.sequentialLetters.test(password)) {
      return "don't use sequences like this!";
    } else {
      return 'moderate';
    }
  } else if (password.length >= 6) {
    if (patterns.repeatingChars.test(password)) {
      return "don't use repeating characters!";
    } else if (patterns.sequentialNumbers.test(password) || patterns.sequentialLetters.test(password)) {
      return "don't use sequences like this!";
    } else {
      return 'weak';
    }
  } else {
    return 'very weak';
  }
}

function displayPasswordStrength(strength) {
  let color = '';
  switch (strength) {
    case "don't use repeating characters!":
    case "don't use sequences like this!":
      color = 'red';
      break;
    case 'very weak':
      color = 'red';
      break;
    case 'weak':
      color = 'orange';
      break;
    case 'moderate':
      color = 'blue';
      break;
    case 'strong':
      color = 'green';
      break;
    case 'very strong':
      color = 'darkgreen';
      break;
    case 'unbreakable':
    color = 'violet';
    break;
     default:
      color = 'black';
      break;
  }

  passwordStrength.textContent = `password strength: ${strength}`;
  passwordStrength.style.color = color;
}

function redirectToRecommended() {
  window.location.href = 'recommended.html';
}

const darkModeToggle = document.getElementById('darkModeToggle');
const body = document.body;

darkModeToggle.addEventListener('click', function() {
  body.classList.toggle('dark-mode');
  const isDarkMode = body.classList.contains('dark-mode');
  localStorage.setItem('darkMode', isDarkMode);
});

function loadDarkModePreference() {
  const isDarkMode = localStorage.getItem('darkMode') === 'true';
  if (isDarkMode) {
    body.classList.add('dark-mode');
  } else {
    body.classList.remove('dark-mode');
  }
}

// Call the function to load the user's dark mode preference on page load
loadDarkModePreference();


let sessionTimeout;

// Function to reset the session timeout
function resetSessionTimeout() {
  clearTimeout(sessionTimeout);
  sessionTimeout = setTimeout(redirectTimeoutPage, 30000); // Timeout duration in milliseconds (30 seconds)
}

// Function to redirect to timeout.html
function redirectTimeoutPage() {
  window.location.href = 'timeout.html';
}

// Event listeners to reset the session timeout on user interaction
document.addEventListener('mousemove', resetSessionTimeout);
document.addEventListener('keypress', resetSessionTimeout);
document.addEventListener('click', resetSessionTimeout);

// Initial call to start the session timeout countdown
resetSessionTimeout();